_docs/Qabel-Protocol-Box.md

```diff
@@ -0,0 +1,580 @@
+1. Insert the share info in the index DM and upload it
+1. Notify the contacts about the new share with a drop message including **dk1**
+
+
+# Handling conflicts
+Conflicts can occour in the small timeframe between checking for changed DM and the
+propagation delay after uploading the updated DM. A client has to check, after a reasonable delay,
+if the changes were overwritten.
+
+1. Download the DM after 10s
+1. Check if the version is the same as in the uploaded file
+1. If a change is detected, repeat the original operation and insert into the DM, set the device, increment the version
+1. Upload the merged DM and repeat.
+
+## Typical scenarios
+
+Client A uploads the DM first, Client B overwrites the DM.
+
+### Non conflicting changes
+
+Scenario: Client A creates a new object and client B creates a new object, the object have different names.
+Solution: Client A inserts the object into the updated DM and uploads it
+
+### Conflicting changes
+
+Scenario: Client A deletes a file, client B changes the file (which uploads a new block)
+Solution: Client A accepts the change from client B
+
+Scenario: Client A changes a file, client B deletes the file
+Solution: Client A inserts the file again
+
+Scenario: Client A changes a file, client B changes the same file
+Solution: Client A inserts the local version of the file with a deconflicted name by adding 'CONFLICT' and the date and time as suffix
+Example: foobar.txt and foobar_CONFLICT_2015-10-23_19:33:23.txt
+
+
+# SQLite Schema
+
+Schema for the SQLite3 database which is used as a directory metadata file (DM). The schema for
+an FM will be included later.
+The JSON documents can be directly translated into this schema.
+
+```SQL
+/*
+The meta table includes the values
+ * root (only needed in the index)
+ * last_change_by
+*/
+CREATE TABLE meta
+(
+       name             VARCHAR(24) PRIMARY KEY,
+       value            TEXT
+);
+
+/*
+A one row table with only the current qabel-box specification version.
+This version is 0 for now and should be checked everytime the databases is opened.
+*/
+CREATE TABLE spec_version
+(
+       version          INTEGER PRIMARY KEY
+);
+
+/*
+Current version and the current time. The time is only used for displaying,
+it should not be trusted.
+* 'id' is meaningless and only for record keeping purposes.
+*/
+CREATE TABLE version
+(
+       id               INTEGER PRIMARY KEY,
+       version          BLOB NOT NULL,
+       time             INTEGER NOT NULL
+);
+
+/*
+Table of all shares in the VOLUME.
+* 'id' is meaningless and only for record keeping purposes.
+* 'ref' is the name of the metadata file
+* 'recipient' is the public key of the recipient
+* 'type' is 0 for a read only share and 1 for a writable share (not implemented yet)
+*/
+CREATE TABLE shares
+(
+       id               INTEGER PRIMARY KEY,
+       ref              VARCHAR(255) NOT NULL,
+       recipient        BLOB NOT NULL,
+       type             INT NOT NULL
```

Comment:
Not is SQLite. LONG, INT, INTEGER and TINYINT are all the same. The names are just for documentation and the drivers.